x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/IpArrayState.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.ByteArray;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.core.Releasables;
+
+/**
+ * Aggregator state for an array of IPs. It is created in a mode where it
+ * won't track the {@code groupId}s that are sent to it and it is the
+ * responsibility of the caller to only fetch values for {@code groupId}s
+ * that it has sent using the {@code selected} parameter when building the
+ * results. This is fine when there are no {@code null} values in the input
+ * data. But once there are null values in the input data it is
+ * <strong>much</strong> more convenient to only send non-null values and
+ * the tracking built into the grouping code can't track that. In that case
+ * call {@link #enableGroupIdTracking} to transition the state into a mode
+ * where it'll track which {@code groupIds} have been written.
+ * <p>
+ *     This class is a specialized version of the {@code X-ArrayState.java.st} template.
+ * </p>
+ */
+public final class IpArrayState extends AbstractArrayState implements GroupingAggregatorState {
+    private static final int IP_LENGTH = 16;
+
+    private final BytesRef scratch = new BytesRef();
+
+    private final byte[] init;
+
+    private ByteArray values;
+
+    IpArrayState(BigArrays bigArrays, BytesRef init) {
+        super(bigArrays);
+        assert init.length == IP_LENGTH;
+        this.values = bigArrays.newByteArray(IP_LENGTH, false);
+        this.init = new byte[IP_LENGTH];
+
+        System.arraycopy(init.bytes, init.offset, this.init, 0, IP_LENGTH);
+        this.values.set(0, this.init, 0, IP_LENGTH);
+    }
+
+    BytesRef get(int groupId) {
```

Comment:
It's a lot more normal to supply the `scratch`. I don't think there's any real harm moving it here, but I think it's less readable. I'd prefer to have it as a parameter just because it's "more normal" even though, I know, it makes the code generation more `if/else`y

Reply:
As the aggregator has only static methods, I would have to create a BytesRef per combine() call. So I'll put this PR in standby and work in porting the @Fixed() annotation to aggs first

Reply:
After checking in depth, well, ther'es no need for the `@Fixed` as we already have a full state object.

Just updated the PR with storing the scratch in there, and added missing tests for the aggregator